### ISSUE TITLE
Sounddevice eats up all my memory

### ISSUE THREAD

**Issue by Plecto123** [spatialaudio/python-sounddevice#58]
Examplecode:

```python
import numpy as np
import time
import sounddevice as sd
 
 
def FindFreq_Resp(f_num, Amp):
    print "Calculating frequency response..."
    Freq=np.logspace(1.3, 4.3, num=f_num)
    for f in Freq:
            xx=np.linspace(0.0, 1, 48000)
            y=Amp*np.sin(f*2.0*np.pi*xx)
            myrecording = sd.playrec(y, 48000, channels=1, blocking=True)        
 
Fs=48000
FindFreq_Resp(100,0.1)
print "Some memory has been eaten"
FindFreq_Resp(100,0.1)
print "Even more memory has been eaten"
FindFreq_Resp(100,0.1)
print "Not much memory left :("
```

If I run the function FindFreq_Resp enough time, the program will run out of swap memory and freeze indefinitely. Why does this happen?

**Comment by mgeier** [comment]
Thanks for the report!

I can confirm that your script keeps growing its memory usage.
When I tried it, it started around 30 MB and reached 250 MB at the end of the script.

I have no clue why this happens ...

Did you try it with Python 3?

When I tried it with Python 3 (after replacing the `print` statements with `print()` function calls), the memory consumption grew in the beginning, but then it stayed slightly below 70 MB.
This is obviously better, but probably it should still be less ...

I don't know if there is anything I can do, any ideas?

**Comment by mgeier** [comment]
@Plecto123 Any news on this?

**Comment by venkat-kittu** [comment]
Hey @Plecto123, how did you solved this problem. Now i am also facing same issue in my case memory is going upto 3.1 GB and program is getting stopped.

**Comment by mgeier** [comment]
There have been some improvements in the latest CFFI release 1.12: https://cffi.readthedocs.io/en/latest/whatsnew.html#v1-12.

The script on the top of this page now still has growing memory usage, but in total it only needed 84 MB with `python2` and about the same with `python3`.

@Plecto123 and @venkat-kittu Does that improve your situation?

**Comment by catchanus** [comment]
I suggest such a solution:

In the 'stop ()' function of 'sounddevice.py' (319th line) add two lines at the end of the 'if' statement:
_last_callback.data = None
_last_callback.out = None

This frees up the memory used by the 'numpy' object.

**Comment by mgeier** [comment]
Thanks @catchanus for the hint, I've implemented it in #250.

Can you (and @Plecto123 and @venkat-kittu) please check if that works for you?

**Comment by catchanus** [comment]
It works perfectly.
